src/core/net/tcp6_ext.cpp

```diff
@@ -0,0 +1,234 @@
+ */
+
+#include "openthread-core-config.h"
+
+#if OPENTHREAD_CONFIG_TCP_ENABLE
+
+#include "tcp6_ext.hpp"
+
+#include "common/as_core_type.hpp"
+#include "common/code_utils.hpp"
+#include "common/error.hpp"
+#include "common/instance.hpp"
+#include "common/locator_getters.hpp"
+#include "common/log.hpp"
+#include "common/random.hpp"
+#include "net/checksum.hpp"
+#include "net/ip6.hpp"
+#include "net/netif.hpp"
+
+#include "../../third_party/tcplp/tcplp.h"
+
+namespace ot {
+namespace Ip6 {
+
+RegisterLogModule("TcpExt");
+
+void TcpCircularSendBuffer::Initialize(void *aDataBuffer, size_t aCapacity)
+{
+    this->mDataBuffer = static_cast<uint8_t *>(aDataBuffer);
+    this->mCapacity   = aCapacity;
+    ForceDiscardAll();
+}
+
+Error TcpCircularSendBuffer::Write(Tcp::Endpoint &aEndpoint,
+                                   const void *   aData,
+                                   size_t         aLength,
+                                   size_t &       aWritten,
+                                   uint32_t       aFlags)
+{
+    Error error = kErrorNone;
+
+    size_t bytesFree = FreeSpace();
+    size_t writeIndex;
+
+    uint32_t flags = 0;
+    size_t   bytesUntilWrap;
+
+    VerifyOrExit(aLength != 0);
+    VerifyOrExit(bytesFree != 0, aLength = 0);
+
+    /*
+     * Handle the case where we don't have enough space to accommodate all of the
+     * provided data.
+     */
+    if (aLength > bytesFree)
+    {
+        aLength = bytesFree;
+    }
+
+    /*
+     * This is a "simplifying" if statement the removes an edge case from the logic
+     * below. It guarantees that a write to an empty buffer will never wrap.
+     */
+    if (this->mCapacityUsed == 0)
+    {
+        this->mStartIndex = 0;
+    }
+
+    writeIndex = GetIndex(this->mStartIndex, this->mCapacityUsed);
+
+    if ((aFlags & OT_TCP_CIRCULAR_SEND_BUFFER_WRITE_MORE_TO_COME) != 0)
+    {
+        flags |= OT_TCP_SEND_MORE_TO_COME;
+    }
+
+    bytesUntilWrap = GetNumBytesUntilWrap(writeIndex);
+    if (aLength <= bytesUntilWrap)
+    {
+        memcpy(&this->mDataBuffer[writeIndex], aData, aLength);
+        if (this->mCapacityUsed == 0 || writeIndex == 0)
+        {
+            /*
+             * this->mCapacityUsed == 0 corresponds to the case where we're writing
+             * to an empty buffer. this->mCapacityUsed != 0 && writeIndex == 0
+             * corresponds to the case where the buffer is not empty and this is
+             * writing the first bytes that wrap.
+             */
+            uint8_t linkIndex;
+            if (this->mCapacityUsed == 0)
+            {
+                linkIndex = this->mFirstSendLinkIndex;
+            }
+            else
+            {
+                linkIndex = 1 - this->mFirstSendLinkIndex;
+            }
+            otLinkedBuffer &dataSendLink = this->mSendLinks[linkIndex];
+            dataSendLink.mData           = &this->mDataBuffer[writeIndex];
+            dataSendLink.mLength         = aLength;
```

Comment:
Here we set the `mData` and `mLength` on `otLinkBuffer`?

How about the  `mNext`? I think we expect to be always `nullptr` right?
Should we set them to null from `Init()` on both `otLinkBuffer` `mSendLinks ` instances?

Reply:
Thanks for catching this. We should set `mNext` every time. We don't use the `SendDoneCallback` so we should not rely on the TCP stack to reset `mNext` to `NULL` when an `otLinkedBuffer` is no longer used.

Reply:
Note that the TCP stack will defensively set `mNext` to `NULL` anyway in the implementation of `SendByReference`, so  none of this really matters. It's still good practice for us to follow our own API, however, and set it to `NULL` every time.

Reply:
Good. Thanks.

> Note that the TCP stack will defensively set mNext to NULL anyway in the implementation of SendByReference, so none of this really matters.

Does it mean we cannot pass an already chained set of `otLinkBuffer` to `SendByReference()`? Only one item?

Reply:
That's correct. You can indicate to the TCP stack not to send out data right away with `OT_TCP_SEND_MORE_TO_COME`. So if you have a bunch of data to send you can just call `SendByReference` multiple times, and pass the `OT_TCP_SEND_MORE_TO_COME` flag on all but the last call.